File: mobile/src/types/difficulty.ts
/**
 * Difficulty rating. Backend stores `difficulty_rating` (text) + `difficulty_score` (number)
 * separately; the wire format on `SudokuPuzzle.difficulty` packs both.
 *
 * `hard` / `impossible` are valid backend values but the mobile app does not
 * expose routes for them yet (mirroring the web app's current state).
 */
export type DifficultyRating = 'beginner' | 'easy' | 'medium' | 'hard' | 'impossible'

export interface Difficulty {
  rating: DifficultyRating
  score?: number
}

File: app/lib/domain/models/difficulty.dart
enum DifficultyRating {
  beginner,
  easy,
  medium,
  hard,
  impossible;

  @override
  String toString() {
    return name.substring(0, 1).toUpperCase() + name.substring(1);
  }

  factory DifficultyRating.fromString(String str) {
    switch (str) {
      case 'Beginner':
        return beginner;
      case 'Easy':
        return easy;
      case 'Medium':
        return medium;
      case 'Hard':
        return hard;
      case 'Impossible':
        return impossible;
      default:
        return beginner;
    }
  }
}
